pkg/controller/service/servicecontroller_test.go

```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
 	}
 }
 
-// TODO(a-robinson): Add tests for update/sync/delete.
+func TestUpdateService(t *testing.T) {
+
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName   string
+		key        string
+		updateFn   func() //Manupulate the structure
+		srv        *v1.Service
+		expectedFn func(error, time.Duration) error //Error comparision function
+	}{
+		{
+			testName: "If updating a valid service",
+			key:      "validKey",
+			srv:      defaultExternalService(),
+			updateFn: func() {
+
+				controller, cloud, _ = newController()
+				controller.cache.getOrCreate("validKey")
+
+			},
+			expectedFn: func(err error, retryDuration time.Duration) error {
+
+				if err != nil {
+					return err
+				}
+				if retryDuration != doNotRetry {
+					return fmt.Errorf("retryDuration Expected=%v Obtained=%v", doNotRetry, retryDuration)
+				}
+				return nil
+			},
+		},
+	}
+
+	for _, tst := range testCases {
+		tst.updateFn()
+		srvCache := controller.cache.getOrCreate(tst.key)
+
+		obtErr, retryDuration := controller.processServiceUpdate(srvCache, tst.srv, tst.key)
+		if err := tst.expectedFn(obtErr, retryDuration); err != nil {
+			t.Errorf("%v processServiceUpdate() %v", tst.testName, err)
+		}
+	}
+
+}
+
+func TestSyncService(t *testing.T) {
+
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName    string
+		key         string
+		updateFn    func() //Function to manipulate the controller element to simulate error
+		expectedErr error  //syncService() only returns error
+	}{
+		{
+			testName: "if an invalid service name is synced",
+			key:      "invalid/key/string",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+
+			},
+			expectedErr: fmt.Errorf("unexpected key format: %q", "invalid/key/string"),
```

Comment:
agreed removed, i was thinking if we have more than once situation where it throws error how do we differentiate it, so far we have only one case that throws error